### PR TITLE
kubeadm: Adds dry-run support for kubeadm using the `--dry-run` option

### LINKED ISSUE #389: Support a `--dry-run` mode for all kubeadm functions
Generally, all POST/PUT/PATCH/DELETEs, etc must be discarded in dry-run mode

On `kubeadm init`, all GET/LISTs will be faked; it will return values that it knows are valid to get the `kubeadm init` flow working well in dry-run mode.
On `kubeadm token`, `join`, `upgrade` where we can access an API server; it will proxy the GET/LIST to that API Server to get real values for kubeadm to react on.

Regarding the filesystem; a temporary directory will be used for writing files. Then the content will be printed and the temp directory removed

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

Adds dry-run support to kubeadm by creating a fake clientset that can get totally fake values (like in the init case), or delegate GETs/LISTs to a real API server but discard all edits like POST/PUT/PATCH/DELETE

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes: https://github.com/kubernetes/kubeadm/issues/389

**Special notes for your reviewer**:

This PR depends on https://github.com/kubernetes/kubernetes/pull/50626, first three commits are from there
This PR is a dependency for https://github.com/kubernetes/kubernetes/pull/48899 (kubeadm upgrades)

I have some small things to fixup and I'll yet write unit tests, but PTAL if you think this is going in the right direction

**Release note**:

```release-note
kubeadm: Adds dry-run support for kubeadm using the `--dry-run` option
```
cc @kubernetes/sig-cluster-lifecycle-pr-reviews @kubernetes/sig-api-machinery-pr-reviews

### CODE REVIEW

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (mattmoyer) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: Is this always correct? I guess it assumes that `kubernetes` is the first allocated Service. That seems roughly reasonable but also possibly a source of confusing bugs since `--dry-run` possibly wouldn't match the real `init` behavior.

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (mattmoyer) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: The comment here is wrong .

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (mattmoyer) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: This comment is wrong as well.

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (mattmoyer) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: This target port might vary per cluster, right?

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (mattmoyer) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: Another option would be to return an obviously fake IP like `0.0.0.0` so that it's clear it's a placeholder.

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (mattmoyer) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: Should we also check against `idr.masterName` here?

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (mattmoyer) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: This comment says "clusterrolebinding" but this is the bootstrap token.

**Review by mattmoyer — CHANGES_REQUESTED** [review]
This will be great to have! It's nice for `init` and `token`, and for debugging during development, but it will be especially nice for `upgrade`.

My only general comment is it needs unit tests. It would also be good to get a review from someone more familiar with the API machinery/client-go code.

File: cmd/kubeadm/app/util/apiclient/wait.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/wait.go]: I don't think so; as it often takes  ~15 secs for control plane to come up
Whether we start polling immediately or after 0.5secs doesn't seem to matter.

Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/wait.go]: But this is cleaner :)

File: staging/src/k8s.io/client-go/testing/actions.go
Comment (luxas) [inline staging/src/k8s.io/client-go/testing/actions.go]: No, more for consistency.
That generic iface and PatchAction were missing somehow...

**Comment by timothysc** [comment]
poke me when the dependent PRs are merged.

**Comment by luxas** [comment]
@timothysc ack, I'll do that.

File: cmd/kubeadm/app/util/apiclient/dryrunclient.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/dryrunclient.go]: I can drop them totally; probably makes sense. What does @timothysc think?

File: cmd/kubeadm/app/util/apiclient/clientbacked_dryrun.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/clientbacked_dryrun.go]: done

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: added a TODO; not for this PR

File: cmd/kubeadm/app/util/apiclient/wait.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/wait.go]: this was in the last PR; it was way too noisy; spitting things out twice every second
Instead this now prints on changed state

Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/wait.go]: But in the last PR, not this one

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: This is the correct behavior and is always the case. It's a special service :)

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: done

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: this doesn't matter; only important thing here is the ClusterIP

File: cmd/kubeadm/app/util/apiclient/init_dryrun.go
Comment (luxas) [inline cmd/kubeadm/app/util/apiclient/init_dryrun.go]: Good idea, now done

**Comment by luxas** [comment]
@timothysc @mattmoyer Rebased and fixed up a small things. Should now work very well.

> My only general comment is it needs unit tests.

I stated above that "I'll yet write unit tests" :smile:
If you have suggestions about what unit tests I should create, please comment on the PR

cc @sttts @ncdc You are familiar with the API machinery, could you take a look and check if this is sane usage of it?

File: cmd/kubeadm/app/cmd/init.go
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: So there are a number of different clientset stubs, I'm wondering if we can dedupe.

File: cmd/kubeadm/app/phases/selfhosting/selfhosting.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/selfhosting/selfhosting.go]: This seems orthogonal to dry-run

**Review by timothysc — CHANGES_REQUESTED** [review]
@ncdc is there a de-facto client-set stub for validating client behavior that you are aware of?

**Comment by ncdc** [comment]
@timothysc could you expand on the exact use case here please?

**Comment by timothysc** [comment]
@ncdc This use case is for a set of stub'd out interactions to the main api.  You basically want a hook mechanism to route some calls to another function.

**Comment by mml** [comment]
cc @caesarxuchao

**Comment by luxas** [comment]
@timothysc @mattmoyer @caesarxuchao @ncdc @sttts This is now rebased. Please take a look and review this. Now unit tests also are in place.

**Comment by caesarxuchao** [comment]
/approve

the client-go changes lgtm.

**Review by timothysc — APPROVED** [review]
/lgtm 

TBH - I don't want to block this, but I'm not a really a fan of much of what is done in here.  The dry-run behavior stubbing is kind of weird.  

Given that it's stub code and not critical path, I'm ok with it.  But it looks like you may be missing some UTs in this patch.  

For now I'll unblock but could you open a follow on issue to address cleanup & re-eval.

**Pull request merged**
